flecsi/util/hpx.hh

```diff
@@ -0,0 +1,136 @@
+/*
+    @@@@@@@@  @@           @@@@@@   @@@@@@@@ @@
+   /@@/////  /@@          @@////@@ @@////// /@@
+   /@@       /@@  @@@@@  @@    // /@@       /@@
+   /@@@@@@@  /@@ @@///@@/@@       /@@@@@@@@@/@@
+   /@@////   /@@/@@@@@@@/@@       ////////@@/@@
+   /@@       /@@/@@//// //@@    @@       /@@/@@
+   /@@       @@@//@@@@@@ //@@@@@@  @@@@@@@@ /@@
+   //       ///  //////   //////  ////////  //
+
+   Copyright (c) 2016, Triad National Security, LLC
+   All rights reserved.
+                                                                              */
+#pragma once
+
+/*! @file */
+
+#include <flecsi-config.h>
+
+#if !defined(FLECSI_ENABLE_HPX)
+#error FLECSI_ENABLE_HPX not defined! This file depends on HPX!
+#endif
+
+#include <hpx/modules/collectives.hpp>
+
+#include <cstddef> // byte
+#include <type_traits>
+#include <vector>
+
+namespace flecsi {
+namespace util {
+namespace hpx {
```

Comment:
Why do we want to have the same utilities but in a different namespace?  Should we just be supplying two different implementations of one function?

Reply:
Hmmm, not sure. I followed the scheme that was in place. I'd be happy to have them in the same namespace and provide different implementations for the two backends. These utilities are used all over the place, so this would simplify the code. Just let me know if you want me to touch the MPI backend at all.

Reply:
Well, `flecsi::util::mpi` isn't actually part of the MPI backend at the moment; it's meant to be used by MPI **tasks** even with the Legion backend, for instance.  However, just like we have some things that are (at the moment) "Legion" or "not Legion", we seem to have several things (like this and `get_num_localities`) that should be "MPI" or "not MPI".  We already have hints at that (`FLECSI_ENABLE_MPI` vs. `FLECSI_RUNTIME_MODEL == FLECSI_RUNTIME_MODEL_mpi`), so it would probably make sense to separate `util/mpi.hh` into a part that was MPI-specific in its **interface** (_e.g._, `type`) and a part that is MPI-specific merely in its **implementation** (like the `one_to_alli` that you reimplemented), given suitable placeholder types for things like `MPI_Comm`.  That latter part would be much like a "partial" backend.

Reply:
Would you be fine if we moved this refactoring into a separate ticket as well?

Reply:
This is an interesting case: it's _new_ code (aside from the comments, which are not quite right anyway), not copied code, so the work to bring it into better alignment with the existing option is more defensibly separable.  That said, there's only one call to any of it (`util::hpx::all_to_allv`, as it happens), and that's in code that is a higher priority to refactor because it does duplicate so much.  If we end up merging the clients of `*::all_to_allv` prior to merging this code, it only makes sense to let that merge be more seamless by giving the implementations the same name (and&mdash;up to trivial differences like `MPI_Comm` _vs._ `hpx::collectives::communicator`&mdash;the same return type!) as well.

Reply:
Yes, I will address in a separate PR.

Reply:
@hkaiser 

Does HPX actually implement these MPI collectives? Or does it actually use MPI? If it the latter, what is the benefit to not just use MPI?

Reply:
The HPX facilities are asynchronous are returning a future to the user. Under the hood (in FleCSI) they use MPI for communication.

Reply:
I think this is up to @tuxfan and @opensdh to decide how much of code duplication are allowed and if it makes sense to do so just to *asynchronize* some utility code that is synchronous even for the Legion backend.

What's the impact to the HPX backend if this file does not exist al all?

Reply:
I have removed the HPX implementations and now rely on the MPI facilities in the HPX backend as well.